GetDKAN/dkan
# Support choosing data dictionaries in dataset node form

Modify json_form_widget to allow selection of Data Dictionaries populated by the metastore.

## Testing steps

1. Build from branch
2. Modify dataset.ui.schema distribution.describedBy to be:

```json
"describedBy": {
  "ui:options": {
    "description": "URL to the data dictionary for the file found at the Download URL.",
      "widget": "list",
      "type": "select",
      "titleProperty": "title",
      "source": {
        "metastoreSchema": "data-dictionary",
        "returnValue": "url"
      }
    }
},
```

3. Create a data dictionary. It can be empty:

```http
POST https://dkan.ddev.site/api/1/metastore/schemas/data-dictionary/items
{
  "identifier": "c4a997e1-e1dc-5139-93b5-0a4e15b012d7",
  "data": {
    "title": "Data dictionary title",
    "fields": []
  }
}
```

4. Try selecting the data dictionary on a new dataset form
5. Check the metastore at https://dkan.ddev.site/api/1/metastore/schemas/dataset/items
6. The describedBy field should be displaying a working URL for the data dictionary.

Note: We really need to write a spec for the ui schema, and possibly make some things about this "source" feature a bit more consistent. 

## Upgrade

This PR changes the data dictionary schema. We did something a little weird with data dictionary where we made a schema with "identifier" and "data" properties but also put the "title" field at the same level. This has caused problems because in different places the metastore expects everything not a dataset to have an identifier/data wrapper. (This whole pattern is problematic but outside the scope of this ticket to fix; we have been [talking about refactoring this for a while](https://github.com/GetDKAN/dkan/issues/3450)).

If you have a /schema folder in your DKAN project outside the DKAN module directory, you'll need to update your data-dictionary.json file with the new schema. If you have existing data dictionary nodes, an update hook will migrate them to the new schema.

To test:

1. Build a site on 2.x branch.
2. Create a data dictionary with the old schema:
```http
POST https://dkan.ddev.site/api/1/metastore/schemas/data-dictionary/items
{
  "identifier": "c4a997e1-e1dc-5139-93b5-0a4e15b012d7",
  "title": "Data dictionary title",
  "data": {
    "fields": []
  }
}
```
3. Checkout `choose-dict` branch
4. Run `drush updb`
5. Make sure you get no errors, and https://dkan.ddev.site/api/1/metastore/schemas/data-dictionary/items shows your data dictionary now with the title under "data".
